feat: astro:env stable (#9160)
withastro/docs: src/content/docs/en/guides/environment-variables.mdx

---
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'

Astro supports 2 main ways for handling environment variables: the `astro:env` module and Vite's built-in support.

> ## `astro:env`

Reviewer: Why did you not just copy the existing docs from the experimental section to start? In the worst case, just drop those in here and then we go from there!

Author: Well I thought it was important to say there are 2 ways to do it for people not using astro:env just yet!

Reviewer: I meant, under the heading specifically for astro:env !  :smile:

## `astro:env`

`astro:env` lets you configure a type-safe schema for your environment variables, and indicate whether they should be available on the server or the client.

To configure a schema, add the `env` option to your Astro config and define your client and server variables:

```js title="astro.config.mjs" ins={4-10}
import { defineConfig, envField } from 'astro/config'

export default defineConfig({
  env: {
    schema: {
      API_URL: envField.string({ context: "client", access: "public", optional: true }),
      PORT: envField.number({ context: "server", access: "public", default: 4321 }),
      API_SECRET: envField.string({ context: "server", access: "secret" }),
    }
  }
})
```

You can import and use your defined variables from the appropriate `/client` or `/server` module:

```astro
---
import { API_URL } from "astro:env/client"
import { API_SECRET_TOKEN } from "astro:env/server"
const data = await fetch(`${API_URL}/users`, {
  method: "GET",
  headers: {
    "Content-Type": "application/json",
    "Authorization": `Bearer ${API_SECRET_TOKEN}`
  },
});
---

<script>
import { API_URL } from "astro:env/client"
fetch(`${API_URL}/ping`)
</script>
```

> **TODO: where to talk about integrations & adapters?**

Reviewer: If you mean for integration/adapter *authors*, then this can be a note with a link to one of the other API pages. I don't think this page should have information about writing integrations.

- **Public client variables**: These variables end up in both your final client and server bundles, and can be accessed from both client and server through the `astro:env/client` module:

   ```js
   import { API_URL } from "astro:env/client"
   ```

- **Public server variables**: These variables end up in your final server bundle and can be accessed on the server through the `astro:env/server` module:

   ```js
   import { PORT } from "astro:env/server"
   ```

- **Secret server variables**: These variables are not part of your final bundle and can be accessed on the server through the `astro:env/server` module:

   ```js
   import { API_SECRET } from "astro:env/server"
   ```

   Worth nothing that by default, secrets are only validated at runtime. To validate secrets on start, check [validateSecrets](#validatesecrets-option).

**Note:** Secret client variables are not supported because there is no safe way to send this data to the client. Therefore, it is not possible to configure both `context: "client"` and `access: "secret"` in your schema.

> ### Data types

Reviewer: A lot of this is starting to get quite API referencey. Not everything is appropriate for a guide page. We should have an `astro:env` section on the API reference page for this content.

Astro gives you access to [Vite's built-in environment variable support](#vites-built-in-support) and includes some [default environment variables for your project](#default-environment-variables) that allow you to access configuration values for your current project (e.g. `site`, `base`), whether your project is running in development or production, and more.

> Astro also provides [a virtual `astro:env` module](/en/reference/configuration-reference/#env) to use and organize your environment variables with type safety. It is available for use inside the Astro context (e.g. Astro components, routes and endpoints; UI framework components; middleware), and [managed with a schema in your Astro configuration](#define-your-schema).

Reviewer: Note, I *intentionally* put the main `#astroenv` link here because we hadn't yet provided a link down to that section! The first link goes to the reference page, but I think people should be taken down to read *about* Astro env.

Open to suggestions if you'd like to change the wording, (it doesn't have to match exactly where the link takes you!) I just wanted to entice people, and let them get directly to that section. :smile:

Author: I feel like having the reference first is a bit odd as it could redirect people before looking at the guide

```suggestion
Astro also provides a way to [use and organize your environment variables with type safety](#type-safe-environment-variables). It is available for use inside the Astro context (e.g. Astro components, routes and endpoints, UI framework components, middleware), and managed with a schema in your Astro configuration. You can see learn more about the `astro:env` module configuration options [in the Configuration reference](/en/reference/configuration-reference/#env).
```

Reviewer: Right, makes sense!

## Type safe environment variables

The `astro:env` API lets you configure a type-safe schema for [environment variables you have set](#setting-environment-variables). This allows you to indicate whether they should be available on the server or the client, and define their data type and additional properties.

<ReadMore>Developing an adapter? See how to [make an adapter compatible with `astro:env`](/en/reference/adapter-reference/#envgetsecret).</ReadMore>

### Basic Usage

#### Define your schema

To configure a schema, add the `env.schema` option to your Astro config:

```js title="astro.config.mjs" ins={4-8}
import { defineConfig } from 'astro/config'

export default defineConfig({
  env: {
    schema: {
      // ...
    }
  }
})
```

> You can then [register variables as a string, number, or boolean](#data-types) using the `envField` helper. Define the [kind of environment variable](#variable-types) by providing a `context` (client or server) and `access` (secret or public) for each variable, and pass any additional properties such as `optional` or `default` in an object:

Author: ```suggestion
You can then [register variables as a string, number, enum or boolean](#data-types) using the `envField` helper. Define the [kind of environment variable](#variable-types) by providing a `context` (client or server) and `access` (secret or public) for each variable, and pass any additional properties such as `optional` or `default` in an object:
```

Reviewer: Yes, good call!

- **Public client variables**: These variables end up in both your final client and server bundles, and can be accessed from both client and server through the `astro:env/client` module:

   ```js
   import { API_URL } from "astro:env/client"
   ```

- **Public server variables**: These variables end up in your final server bundle and can be accessed on the server through the `astro:env/server` module:

   ```js
   import { PORT } from "astro:env/server"
   ```

- **Secret server variables**: These variables are not part of your final bundle and can be accessed on the server through the `astro:env/server` module:

   ```js
   import { API_SECRET } from "astro:env/server"
   ```

   By default, secrets are only validated at runtime. You can enable validating private variables on start by [configuring `validateSecrets: true`](/en/reference/configuration-reference/#envvalidatesecrets).

:::note
**Secret client variables** are not supported because there is no safe way to send this data to the client. Therefore, it is not possible to configure both `context: "client"` and `access: "secret"` in your schema.
:::

### Data types

> There are currently four data types supported: strings, numbers, booleans and enums:

Author: there are not required, it's just the most common props. It was a middle ground between showing nothing (no codeblock) and showing every possible option

> There are currently four data types supported: strings, numbers, booleans and enums:

Author: So no change required here isn't it?

```js
import { envField } from "astro/config"

envField.string({
   // context & access
   optional: true,
   default: "foo",
})

envField.number({
   // context & access
   optional: true,
   default: 15,
})

envField.boolean({
   // context & access
   optional: true,
   default: true,
})

envField.enum({
   // context & access
   values: ['foo', 'bar', 'baz'],
   optional: true,
   default: 'baz',
})
```

> Some of them come with more optional validation rules, check out the [source](https://github.com/withastro/astro/blob/main/packages/astro/src/env/schema.ts#L3-L45) to learn more.

Reviewer: Is there a reason we are not *documenting* these? I would rather point people to a docs link than source code.

Author: Because we have no good place for it! It's not a runtime API thing, it's a configuration reference thing but that one is fully automated. IMO most people will discover these options with autocompletion in their editor

Reviewer: Right, so the docs should be in config reference!

Reviewer: I suggest we make a dummy `envField` entry in config reference locally in this branch, then go make one properly in the types/public file to fill it in.

> Some of them come with more optional validation rules, check out the [source](https://github.com/withastro/astro/blob/main/packages/astro/src/env/schema.ts#L3-L45) to learn more.

Reviewer: ```suggestion
<ReadMore>For a complete list of validation fields, see the [`envField` API reference](/en/reference/configuration-reference/#envschema).</ReadMore>
```
